pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,100 @@
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/initialization"
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+const (
+	initializedCondition conditionsv1alpha1.ConditionType = "APIBindingsInitialized"
+
+	failedBootstrappingReason = "Failed to bootstrap API Bindings."
+)
+
+func (b *APIBinder) reconcile(ctx context.Context, clusterWorkspace *tenancyv1alpha1.ClusterWorkspace) {
+	logger := logging.WithObject(klog.FromContext(ctx), clusterWorkspace)
+	ctx = klog.NewContext(ctx, logger)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("Initializing APIBindings")
+	bindingClient := b.kcpClusterClient.Cluster(clusterName).ApisV1alpha1().APIBindings()
+	cwtClient := b.kcpClusterClient.Cluster(clusterName).TenancyV1alpha1().ClusterWorkspaceTypes()
+
+	cwtName := string(clusterWorkspace.Spec.Type.Name)
+	cwt, err := cwtClient.Get(ctx, cwtName, metav1.GetOptions{})
+	if err != nil {
+		errors = append(errors, fmt.Errorf("could not find ClusterWorkspaceType: %q, %w", cwtName, err))
+	}
+
+	for i, exportRef := range cwt.Spec.APIBindings {
+		if exportRef.Reference.WorkspacePath == "" {
+			errors = append(errors, fmt.Errorf("spec.bindings[%d] invalid: no workspace reference", i))
+			continue
+		}
+		hash := sha256.Sum224([]byte(exportRef.Reference.WorkspacePath + exportRef.Reference.Name))
+		name := fmt.Sprintf("%x", hash)
+		logger.V(2).Info("Creating APIBinding", "binding", name)
+		_, err := bindingClient.Create(ctx, &apisv1alpha1.APIBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: apisv1alpha1.APIBindingSpec{
+				Reference: apisv1alpha1.ExportReference{
+					Workspace: &apisv1alpha1.WorkspaceExportReference{
+						Path:       exportRef.Reference.WorkspacePath,
+						ExportName: exportRef.Reference.Name,
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
+			errors = append(errors, fmt.Errorf("could not create APIBinding %q: %w", name, err))
+			continue
+		}
+	}
+	if len(errors) > 0 {
+		logger.V(2).Info("Failed initializing APIBindings")
+		conditions.MarkFalse(
+			clusterWorkspace,
+			initializedCondition,
+			failedBootstrappingReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"encountered errors: %v",
+			utilerrors.NewAggregate(errors),
+		)
+	} else {
+		logger.V(2).Info("Succeeded initializing APIBindings")
```

Comment:
We need to make sure the APIs are actually bound, or we break the expectation that when their workspace exits initializing, it's ready to use

Reply:
That's covered by another controller (phase change to ready) right?

Reply:
I don't think so, no ... you are saying a controller updates the state of a `ClusterWorkspace` when `APIBinding`s inside of it are not bound? Not to my knowledge.

Reply:
Yes, we have one that transitions the phase to ready only when the bindings are good to go

Reply:
Huh, wow, that is even before any initializer work was done. IMHO honestly we should not aim to hard-code this concept across N controllers. What's the reason that we shouldn't define *this* controller's work as a) creating the bindings and b) ensuring they are bound before it removes the iniitalizer? I assume we won't have call-outs for any other initializers in the CW reconciler.

Reply:
Clearly my brain is holding onto nothing ... [here](https://github.com/kcp-dev/kcp/pull/1316#discussion_r903684649) we were having the same comment months ago.

Reply:
I think you're correct in that we need to wait for the `APIBindings` to be successfully bound, not just that we submitted a create request and got no error.